lockfile/src/python.rs

```diff
@@ -34,11 +32,19 @@ impl Parse for PyRequirements {
     }
 
     fn is_path_lockfile(&self, path: &Path) -> bool {
-        path.file_name() == Some(OsStr::new("requirements.txt"))
+        let file_name = match path.file_name().and_then(|f| f.to_str()) {
+            Some(f) => f,
+            None => return false,
+        };
+        file_name.starts_with("requirements") && file_name.ends_with(".txt")
     }
 
     fn is_path_manifest(&self, path: &Path) -> bool {
-        path.file_name() == Some(OsStr::new("requirements.txt"))
+        path.file_name() == Some(OsStr::new("requirements.in"))
+            || path.file_name() == Some(OsStr::new("pyproject.toml"))
+            || path.file_name() == Some(OsStr::new("setup.cfg"))
+            || path.file_name() == Some(OsStr::new("setup.py"))
```

Comment:
Testing with the `pyproject.toml` file from `phylum-ci` resulted in a failure. An initial test of a sample `setup.py` also failed. `setup.cfg` has not been tested yet. More detail will be provided next week.

Reply:
## `pyproject.toml`

Testing again with `pyproject.toml` files succeeded this time...but only after some trial and error. Things get squirrelly when the `pyproject.toml` file is taken out of context of the project files it is meant to cover. Initially, a lot of errors of the sort described [in this issue](https://github.com/jazzband/pip-tools/issues/1711) were seen. This was true for both a "lifted" `pyproject.toml` file (taken from `phylum-ci` and dropped in a different directory) and for a freshly created file using `poetry init` and specifying a few main and dev dependencies. Instead of trying to find the root cause of these failures, it was finally discovered that testing _within_ a true poetry project...in the root directory...worked as intended. Well, it worked in the sense that the `-t poetry` produced the correct results. Running with `-t pip` resulted in a lockfile, but with just a few differences in the resolved dependency graph:

<details><summary>Details</summary>
<p>

```
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip pyproject.toml
Generating lockfile for manifest "pyproject.toml" using Pip…
[
  {
    "name": "certifi",
    "version": "2023.5.7",
    "type": "pypi"
  },
  {
    "name": "cffi",
    "version": "1.15.1",
    "type": "pypi"
  },
  {
    "name": "charset-normalizer",
    "version": "3.1.0",
    "type": "pypi"
  },
  {
    "name": "cryptography",
    "version": "40.0.2",
    "type": "pypi"
  },
  {
    "name": "idna",
    "version": "3.4",
    "type": "pypi"
  },
  {
    "name": "markdown-it-py",
    "version": "2.2.0",
    "type": "pypi"
  },
  {
    "name": "mdurl",
    "version": "0.1.2",
    "type": "pypi"
  },
  {
    "name": "packaging",
    "version": "23.1",
    "type": "pypi"
  },
  {
    "name": "pathspec",
    "version": "0.11.1",
    "type": "pypi"
  },
  {
    "name": "pycparser",
    "version": "2.21",
    "type": "pypi"
  },
  {
    "name": "pygments",
    "version": "2.15.1",
    "type": "pypi"
  },
  {
    "name": "requests",
    "version": "2.30.0",
    "type": "pypi"
  },
  {
    "name": "rich",
    "version": "13.3.5",
    "type": "pypi"
  },
  {
    "name": "ruamel-yaml",
    "version": "0.17.24",
    "type": "pypi"
  },
  {
    "name": "ruamel-yaml-clib",
    "version": "0.2.7",
    "type": "pypi"
  },
  {
    "name": "urllib3",
    "version": "2.0.2",
    "type": "pypi"
  }
]

❯ poetry show
certifi            2023.5.7 Python package for providing Mozilla's CA Bundle.
cffi               1.15.1   Foreign Function Interface for Python calling C code.
charset-normalizer 3.1.0    The Real First Universal Charset Detector. Open, modern and actively maintained alternative to Chardet.
commonmark         0.9.1    Python parser for the CommonMark Markdown spec
cryptography       40.0.2   cryptography is a package which provides cryptographic recipes and primitives to Python developers.
idna               3.4      Internationalized Domain Names in Applications (IDNA)
packaging          23.1     Core utilities for Python packages
pathspec           0.11.1   Utility library for gitignore style pattern matching of file paths.
pycparser          2.21     C parser in Python
pygments           2.15.1   Pygments is a syntax highlighting package written in Python.
requests           2.30.0   Python HTTP for Humans.
rich               12.6.0   Render rich text, tables, progress bars, syntax highlighting, markdown and more to the terminal
ruamel.yaml        0.17.24  ruamel.yaml is a YAML parser/emitter that supports roundtrip preservation of comments, seq/map flow style, and map key order
ruamel.yaml.clib   0.2.7    C version of reader, parser and emitter for ruamel.yaml derived from libyaml
urllib3            2.0.2    HTTP library with thread-safe connection pooling, file post, and more.
```

The difference is in `rich`, `commonmark`, `mdurl`, and `markdown-it-py`
</p>
</details> 

The other difference is that `-t pip` resulted in a lot fewer dependencies than `-t poetry`. This is presumed to be the difference between resolving just the `main` deps (`pip-compile`) and all the deps (`poetry`).

---

## `requirements.in`

Testing here showed that `phylum parse`, with a valid `.phylum_project` file, worked as intended. However, using the same file but with a direct `phylum parse -t pip <path/to/requirements.in>` failed:

<details><summary>Details</summary>
<p>

```
~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❌ 1 ❯ ~/dev/phylum/localdev/cli/target/debug/phylum --version
phylum v5.1.0-9-g6dcf572

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ l
total 16
drwxr-xr-x   8 maxrake  staff   256B May  8 20:31 .
drwxr-xr-x  25 maxrake  staff   800B May  5 13:59 ..
-rw-------   1 maxrake  staff   176B May  8 20:31 .phylum_project
drwxr-xr-x   5 maxrake  staff   160B May  8 20:32 pip
drwxr-xr-x   4 maxrake  staff   128B May  8 18:06 poetry_init
-rw-r--r--   1 maxrake  staff   215B May  8 20:30 requirements-locked.txt
drwxr-xr-x   8 maxrake  staff   256B May  8 16:25 yarn
drwxr-xr-x  11 maxrake  staff   352B May  8 16:27 yarn_init

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ l pip
total 32
drwxr-xr-x  5 maxrake  staff   160B May  8 20:32 .
drwxr-xr-x  8 maxrake  staff   256B May  8 20:31 ..
-rw-r--r--  1 maxrake  staff   5.2K May  5 18:21 pyproject.toml
-rw-r--r--  1 maxrake  staff     8B May  5 13:59 requirements.in
-rw-r--r--@ 1 maxrake  staff   604B May  5 18:22 setup.py

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ cat pip/requirements.in
pyyaml


~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ cat .phylum_project
id: afb35f8b-bce5-41ce-82a1-dee88e1bf39d
name: manifest_test
created_at: 2023-05-05T14:01:30.038872-05:00
group_name: null
lockfiles:
- path: ./pip/requirements.in
  type: pip

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse
Generating lockfile for manifest "/Users/maxrake/dev/delme/manifest_test/./pip/requirements.in" using Pip…
[
  {
    "name": "pyyaml",
    "version": "6.0",
    "type": "pypi"
  }
]

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip pip/requirements.in
Generating lockfile for manifest "pip/requirements.in" using Pip…
❗ Error: package manager exited with non-zero status

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❌ 1 ❯
```

</p>
</details> 

---

## `setup.py`

Testing with a `setup.py` file taken out of context (i.e., lifted from another project and dropped into a directory without the rest of the project) resulted in an error:

<details><summary>Details</summary>
<p>

```
~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme) took 8s
❯ cat .phylum_project
id: afb35f8b-bce5-41ce-82a1-dee88e1bf39d
name: manifest_test
created_at: 2023-05-05T14:01:30.038872-05:00
group_name: null
lockfiles:
- path: ./pip/setup.py
  type: pip

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ cat pip/setup.py
from __future__ import annotations

from setuptools import setup

setup(
    name="fake_with_deps",
    version=0.1,
    install_requires=[
        "python-dateutil>=2.4.2,<2.5",
        "colorama<0.4.0,>=0.3.7",
        "cornice<1.1,>=1.0.0",
        "enum34<1.1.7,>=1.0.4",
        "six>1.5,<=1.8",
        "ipaddress<1.1,>=1.0.16",
        "jsonschema<3.0,>=2.4.0",
        "pyramid<1.6,>=1.5.7",
        "pyzmq<14.8,>=14.7.0",
        "simplejson>=3.5,!=3.8,>3.9",
        "SQLAlchemy!=0.9.5,<2.0.0,>=0.7.8,>=1.0.0",
        "python-memcached>=1.57,<2.0",
        "xmltodict<=0.11,>=0.4.6",
    ],
)

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse
Generating lockfile for manifest "/Users/maxrake/dev/delme/manifest_test/./pip/setup.py" using Pip…
❗ Error: package manager exited with non-zero status

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme) took 3s
❌ 1 ❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip pip/setup.py
Generating lockfile for manifest "pip/setup.py" using Pip…
❗ Error: package manager exited with non-zero status

~/dev/delme/manifest_test  4  2 via 🐍 v3.11.3 (delme)
❌ 1 ❯
```

</p>
</details>

Testing with the file in context resulted in success:

<details><summary>Details</summary>
<p>

```
~/Downloads/Levenshtein-0.21.0  4  9 via △ v3.26.3 via 🐍 v3.11.3 took 3s
❯ l
total 112
drwxr-xr-x@ 15 maxrake  staff   480B Apr 19 16:44 .
drwx------+ 36 maxrake  staff   1.1K May  8 10:48 ..
-rw-r--r--@  1 maxrake  staff   1.8K Apr 19 16:44 CMakeLists.txt
-rw-r--r--@  1 maxrake  staff    18K Apr 19 16:44 COPYING
-rw-r--r--@  1 maxrake  staff   7.2K Apr 19 16:44 HISTORY.md
-rw-r--r--@  1 maxrake  staff   476B Apr 19 16:44 MANIFEST.in
-rw-r--r--@  1 maxrake  staff   3.3K Apr 19 16:44 PKG-INFO
-rw-r--r--@  1 maxrake  staff   2.5K Apr 19 16:44 README.md
drwxr-xr-x@  3 maxrake  staff    96B Apr 19 16:44 _custom_build
drwxr-xr-x@  3 maxrake  staff    96B Apr 19 16:44 extern
-rw-r--r--@  1 maxrake  staff   134B Apr 19 16:44 pyproject.toml
-rw-r--r--@  1 maxrake  staff    38B Apr 19 16:44 setup.cfg
-rw-r--r--@  1 maxrake  staff   1.1K Apr 19 16:44 setup.py
drwxr-xr-x@  4 maxrake  staff   128B Apr 19 16:44 src
drwxr-xr-x@  7 maxrake  staff   224B Apr 19 16:44 tests

~/Downloads/Levenshtein-0.21.0  4  9 via △ v3.26.3 via 🐍 v3.11.3
❯ cat setup.py
from skbuild import setup

with open('README.md', 'rt', encoding="utf8") as f:
    readme = f.read()

setup(
    name="Levenshtein",
    version="0.21.0",
    url="https://github.com/maxbachmann/Levenshtein",
    author="Max Bachmann",
    install_requires=["rapidfuzz >= 2.3.0, < 4.0.0"],
    author_email="contact@maxbachmann.de",
    description="Python extension for computing string edit distances and similarities.",
    long_description=readme,
    long_description_content_type="text/markdown",

    license="GPL",
    license_file = "COPYING",
    classifiers=[
        "Programming Language :: Python :: 3",
        "Programming Language :: Python :: 3.6",
        "Programming Language :: Python :: 3.7",
        "Programming Language :: Python :: 3.8",
        "Programming Language :: Python :: 3.9",
        "Programming Language :: Python :: 3.10",
        "Programming Language :: Python :: 3.11",
        "License :: OSI Approved :: GNU General Public License v2 or later (GPLv2+)"
    ],

    packages=["Levenshtein"],
    package_dir={'':'src'},
    package_data={
        "Levenshtein": ["*.pyi", "py.typed"]
    },
    python_requires=">=3.6"
)

~/Downloads/Levenshtein-0.21.0  4  9 via △ v3.26.3 via 🐍 v3.11.3
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip setup.py
Generating lockfile for manifest "setup.py" using Pip…
[
  {
    "name": "rapidfuzz",
    "version": "3.0.0",
    "type": "pypi"
  }
]
```

</p>
</details>

## `setup.cfg`

Testing with a `setup.cfg` file taken out of context (i.e., lifted from another project and dropped into a directory without the rest of the project) resulted in the same type of error as was seen for `setup.py`. Testing with it in context worked just fine:

<details><summary>Details</summary>
<p>

```
~/Downloads/pronto-2.5.4  2  7 via 🐍 v3.11.3
❌ 1 ❯ l
total 112
drwxr-xr-x@ 11 maxrake  staff   352B Apr 10 12:49 .
drwx------+ 38 maxrake  staff   1.2K May  8 20:47 ..
-rw-r--r--@  1 maxrake  staff    17K Apr 10 12:49 CHANGELOG.md
-rw-r--r--@  1 maxrake  staff   1.1K Apr 10 12:49 COPYING
-rw-r--r--@  1 maxrake  staff   134B Apr 10 12:49 MANIFEST.in
-rw-r--r--@  1 maxrake  staff   9.2K Apr 10 12:49 PKG-INFO
-rw-r--r--@  1 maxrake  staff   7.7K Apr 10 12:49 README.md
drwxr-xr-x@ 17 maxrake  staff   544B Apr 10 12:49 pronto
drwxr-xr-x@  8 maxrake  staff   256B Apr 10 12:49 pronto.egg-info
-rw-r--r--@  1 maxrake  staff   2.6K Apr 10 12:49 setup.cfg
-rw-r--r--@  1 maxrake  staff    60B Apr 10 12:49 setup.py

~/Downloads/pronto-2.5.4  2  7 via 🐍 v3.11.3
❯ cat setup.py
#!/usr/bin/env python

import setuptools
setuptools.setup()

~/Downloads/pronto-2.5.4  2  7 via 🐍 v3.11.3
❯ cat setup.cfg
[metadata]
name = pronto
version = attr: pronto.__version__
author = Martin Larralde
author-email = martin.larralde@embl.de
home-page = https://github.com/althonos/pronto
description = Python frontend to ontologies.
long-description = file: README.md
long_description_content_type = text/markdown
license = MIT
license-file = COPYING
platform = any
keywords = bioinformatics, ontology, OBO, OWL, convert, parse
classifiers =
	Development Status :: 4 - Beta
	Programming Language :: Python
	Programming Language :: Python :: 3
	Programming Language :: Python :: 3.7
	Programming Language :: Python :: 3.8
	Programming Language :: Python :: 3.9
	Programming Language :: Python :: 3.10
	Programming Language :: Python :: 3.11
	Intended Audience :: Developers
	Intended Audience :: Science/Research
	Intended Audience :: Healthcare Industry
	License :: OSI Approved :: MIT License
	Topic :: Scientific/Engineering :: Bio-Informatics
	Topic :: Software Development :: Libraries :: Python Modules
	Operating System :: OS Independent
	Typing :: Typed
project_urls =
	Bug Tracker = https://github.com/althonos/pronto/issues
	Changelog = https://pronto.readthedocs.io/en/latest/changes.html
	Documentation = https://pronto.readthedocs.io/en/latest/
	Coverage = https://app.codecov.io/gh/althonos/pronto
	CI = https://github.com/althonos/pronto/actions

[options]
zip_safe = true
include_package_data = true
python_requires = >= 3.7
packages =
	pronto
	pronto.entity
	pronto.logic
	pronto.utils
	pronto.parsers
	pronto.serializers
test_suite = tests
setup_requires =
	setuptools >=46.4
install_requires =
	chardet ~=5.0
	fastobo ~=0.12.2
	networkx >=2.3,<4.0
	python-dateutil ~=2.8

[options.package_data]
pronto = py.typed

[sdist]
formats = zip

[bdist_wheel]
universal = 1

[coverage:report]
exclude_lines =
	pragma: no cover
	raise NotImplementedError
	if __name__ == .__main__.:
	except ImportError:
	if typing.TYPE_CHECKING:
	@abc.abstractmethod

[coverage:run]
branch = True
source = pronto

[green]
file-pattern = test_*.py
verbose = 2
no-skip-report = true
quiet-stdout = true
run-coverage = true

[pydocstyle]
inherit = false
ignore = D100,D101,D102,D105,D107,D200,D203,D213,D406,D407
match-dir = (?!tests)(?!resources)(?!docs)(?!build)[^\.].*
match = (?!test)(?!setup)[^\._].*\.py

[mypy]
ignore_missing_imports = true

[mypy-pronto.*]
disallow_any_decorated = false
disallow_any_generics = false
disallow_any_unimported = false
disallow_subclassing_any = true
disallow_untyped_calls = false
disallow_untyped_defs = false
ignore_missing_imports = false
warn_unused_ignores = false
warn_return_any = false

[egg_info]
tag_build =
tag_date = 0


~/Downloads/pronto-2.5.4  2  7 via 🐍 v3.11.3
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip setup.cfg
Generating lockfile for manifest "setup.cfg" using Pip…
[
  {
    "name": "chardet",
    "version": "5.1.0",
    "type": "pypi"
  },
  {
    "name": "fastobo",
    "version": "0.12.2",
    "type": "pypi"
  },
  {
    "name": "networkx",
    "version": "3.1",
    "type": "pypi"
  },
  {
    "name": "python-dateutil",
    "version": "2.8.2",
    "type": "pypi"
  },
  {
    "name": "six",
    "version": "1.16.0",
    "type": "pypi"
  }
]
```

</p>
</details>

Reply:
We don't intend to support these files when taken out of context... The context is important.

The `requirements.in` issue needs further investigation.

Reply:
> The `requirements.in` issue needs further investigation.

Maybe it has something to do with sub-directories, canonicalization, or otherwise providing an unexpected value to the spawned `pip-compile` command. Here is more detailed output showing that the command works when applied to a manifest in the same directory but not when the manifest is in a different directory. Using `pip-compile` directly works in both cases.

<details><summary>Details</summary>
<p>

```
~/dev/delme/manifest_test  4  1 via 🐍 v3.11.3 (delme)
❯ ~/dev/phylum/localdev/cli/target/debug/phylum --version
phylum v5.1.0-9-g6dcf572

~/dev/delme/manifest_test  4  1 via 🐍 v3.11.3 (delme)
❯ ls -alh
total 8
drwxr-xr-x   7 maxrake  staff   224B May  9 16:00 .
drwxr-xr-x  25 maxrake  staff   800B May  5 13:59 ..
-rw-------   1 maxrake  staff   176B May  9 15:28 .phylum_project
drwxr-xr-x   6 maxrake  staff   192B May  9 15:58 pip
drwxr-xr-x   4 maxrake  staff   128B May  8 18:06 poetry_init
drwxr-xr-x   8 maxrake  staff   256B May  8 16:25 yarn
drwxr-xr-x  11 maxrake  staff   352B May  8 16:27 yarn_init

~/dev/delme/manifest_test  4  1 via 🐍 v3.11.3 (delme)
❯ ls -alh pip
total 40
drwxr-xr-x  6 maxrake  staff   192B May  9 15:58 .
drwxr-xr-x  7 maxrake  staff   224B May  9 16:00 ..
-rw-r--r--  1 maxrake  staff   5.2K May  5 18:21 pyproject.toml
-rw-r--r--  1 maxrake  staff     8B May  5 13:59 requirements.in
-rw-r--r--@ 1 maxrake  staff   2.6K May  8 20:48 setup.cfg
-rw-r--r--@ 1 maxrake  staff   604B May  5 18:22 setup.py

~/dev/delme/manifest_test  4  1 via 🐍 v3.11.3 (delme)
❯ cat pip/requirements.in
pyyaml


~/dev/delme/manifest_test  4  1 via 🐍 v3.11.3 (delme)
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip pip/requirements.in
Generating lockfile for manifest "pip/requirements.in" using Pip…
❗ Error: package manager exited with non-zero status

~/dev/delme/manifest_test  4  1 via 🐍 v3.11.3 (delme)
❌ 1 ❯ deactivate

~/dev/delme/manifest_test  4  1
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip pip/requirements.in
Generating lockfile for manifest "pip/requirements.in" using Pip…
❗ Error: package manager exited with non-zero status

~/dev/delme/manifest_test  4  1
❌ 1 ❯ cd pip

~/dev/delme/manifest_test/pip  0  4 is 📦 v0.28.1 via 🐍 v3.11.3
❯ ~/dev/phylum/localdev/cli/target/debug/phylum parse -t pip requirements.in
Generating lockfile for manifest "requirements.in" using Pip…
[
  {
    "name": "pyyaml",
    "version": "6.0",
    "type": "pypi"
  }
]

~/dev/delme/manifest_test/pip  0  4 is 📦 v0.28.1 via 🐍 v3.11.3
❯ cd ..

~/dev/delme/manifest_test  4  1
❯ pip-compile -o requirements-locked.txt -v pip/requirements.in
WARNING: the legacy dependency resolver is deprecated and will be removed in future versions of pip-tools. The default resolver will be changed to 'backtracking' in pip-tools 7.0.0. Specify --resolver=backtracking to silence this warning.
Using indexes:
  https://pypi.org/simple

                          ROUND 1
Current constraints:
  pyyaml (from -r pip/requirements.in (line 1))

Finding the best candidates:
  found candidate pyyaml==6.0 (constraint was <any>)

Finding secondary dependencies:
  pyyaml==6.0               requires -
------------------------------------------------------------
Result of round 1: stable, done

#
# This file is autogenerated by pip-compile with Python 3.11
# by the following command:
#
#    pip-compile --output-file=requirements-locked.txt pip/requirements.in
#
pyyaml==6.0
    # via -r pip/requirements.in
```

</p>
</details>

Reply:
> We don't intend to support these files when taken out of context... The context is important.

Agreed. However, users do all sorts of odd things. The likelihood that they will attempt to parse one of these files out of context is high. The suggestion is to provide hints or clear advice where it will be seen. That could mean more detailed error messages, a FAQ, pre-requisites in the documentation, etc. Even if not supported, the lack of support should be documented somehow.

This is also helpful for selfish reasons: a week+ from now, when someone shows up with an `❗ Error: package manager exited with non-zero status` message asking for help we'll have a better clue as to what is going on.

Reply:
Yep. That's the goal of #1085 

From our perspective, it's impossible to tell if the user has taken the file out of context or not. But the underlying tool should know and should (hopefully) provide a decent error message. We should pass that along.